### PR TITLE
feat: add Nel Applicative instance

### PR DESCRIPTION
**Pull request by stephentetley**
This PR adds an Applicative instance to Nel. This allows the Monad instance to be uncommented.

Nel already had `singleton` for `point` so I've just added tests for `ap`.

### CODE REVIEW

File: main/src/library/Nel.flix [hunk 1]
```diff
@@ -220,6 +224,27 @@ namespace Nel {
...
+    ///
+    pub def ap(f: Nel[a -> b & ef], l: Nel[a]) : Nel[b] & ef =  match f {
+        case Nel(f1, Nil) => map(f1, l)
+        case Nel(f1, rs)  => apHelperList(f1, rs, l, ks -> ks as & ef)
+    }
+
+    ///
+    /// Helper function for `ap`.
+    ///
+    def apHelperList(f: a -> b & ef, fs: List[a -> b & ef], l: Nel[a], k: Nel[b] -> Nel[b] & ef): Nel[b] & ef = match fs {
```
Comment (magnus-madsen) [inline main/src/library/Nel.flix:240]: Since we now have local defs (let-rec) the goal is to have all of these helpers inlined.

**Review by magnus-madsen — COMMENTED** [review]
Looks good. Only one comment about using local defs.

**Comment by stephentetley** [comment]
Thanks Magnus - I move to a local definition.

Because the helper was fairly complicated I wanted the help of a type signature to develop/debug it.

**Comment by magnus-madsen** [comment]
> Thanks Magnus - I move to a local definition.
> 
> Because the helper was fairly complicated I wanted the help of a type signature to develop/debug it.

OK. Note that local defs can still have type signatures (just not the return type atm)

**Comment by stephentetley** [comment]
I've just push an update - I didn't use an internal type signature but I did add an internal comment as the loop uses a worker wrapper idiom to avoid matching on the empty list.

Maybe it is just the observer effect but the library test suite seemed to take a lot longer to run.

File: main/src/library/Nel.flix (hunk 1 above)
Comment (stephentetley) [reply main/src/library/Nel.flix]: Updated to use loop

**Comment by magnus-madsen** [comment]
> I've just push an update - I didn't use an internal type signature but I did add an internal comment as the loop uses a worker wrapper idiom to avoid matching on the empty list.
> 
> Maybe it is just the observer effect but the library test suite seemed to take a lot longer to run.

Could be a blow-up in Boolean formulas. Ill check locally.

**Comment by magnus-madsen** [comment]
A few things:

- Whether a Boolean formula "blows-up" is never the fault of the Flix programmer. It means we have to do better in the compiler. (In fact we are already working on it with several ideas).
- I do see a very slightly complex formula:

image

But I am surprised that this would have any impact on running time. Can you actually test a difference if you run it with a stop watch? (I can also try when I get to my desktop in a few days).

**Comment by magnus-madsen** [comment]
Indeed the GitHub build is 2min slower on this PR! (If those numbers can be trusted).

**Comment by magnus-madsen** [comment]
My guess is that this slow down only happens with the local def. Can we confirm that?

**Comment by stephentetley** [comment]
I get these results (2 runs each)

with loop1 - 2:28
with loop2 - 2:22

helper-fun1 - 2:20
helper-fun2 - 2:24

fresh-sandbox1 - 2:31
fresh-sandbox2 - 2:19


fresh sandbox doesn't have Nel.ap defined and doesn't have its 5 tests.

**Comment by magnus-madsen** [comment]
Seems within the margin of error.

**Pull request merged**